README.md
16cc
----

The 16candles compiler

Joe Jevnik, Ted Meyer, and Jack Pugmire
Licensed under the GNU Public License version 2.


Building and Installing 16candles
---------------------------------

To build 16candles, the host machine must have CMake and a C compiler
supporting the gnu11 standard, and must be somewhat POSIX-compliant. Virtually
all Unix-like operating systems, including GNU/Linux, FreeBSD, and Mac OS X are
capable of compiling 16c.

The fastest way to build 16candles is to simply enter these commands from the
project directory:

    $ cmake .
    $ make

It is recommending that you create a build directory. With this in mind, the
commands become:

    $ mkdir build
    $ cd build
    $ cmake ..
    $ make .

There are no installation targets for the build system yet. In the meantime,
feel free to move or copy the created executables manually and pretend that a
build script is doing it.


Additional Projects
-------------------

- The VM: [https://github.com/16candles/16c]
- Compiler written in Haskell (supports when/unless):
  [h16cc](https://github.com/llllllllll/16candles_haskell)
- Debugger: [16db](https://github.com/16candles/16db)


ChangeLog
2014-02-03 Jack Pugmire <comradetechnology@gmail.com>
	* ChangeLog: added ChangeLog
	* src/common: added src/common directory
	* src/common/arch.h: moved arch.h to common

2014-02-04 Joe Jevnik <joejev@gmail.com>
	* src/common/arch.h: Prefixed register bytes with OP_, added more
	typedefs
	* src/compiler/processor.c: Started the processing commands
	* src/compiler/processor.h: Added and started
	* src/compiler/operations.c: Added the binary, comparison, and unary
	operators
	* src/compiler/operations.h: Added types for the different kinds of
	operators as function pointers
	* src/compiler/memory.c: Added a simple test memory model
	* src/compiler/memory.h: Added a simple test memory model
	* src/compiler/makefile: added src/compiler/makefile
	* examples/bin_op_test.16c: Added test file
	* examples/cmp_op_test.16c: Added test file
	* examples/goodbyeworld.16c: Fixed ascii codes
	* TODO: added TODO

2014-02-04 Jack Pugmire <comradetechnology@gmail.com>
	* src/machine/memory.c, src/machine/memory.h: rewrote memory
	infrastructure using a memory-mapped file to represent virtual
	memory.
	* src/machine/processor.c, src/machine/processor.h: changed
	processor to work with the new memory system
	* src/machine/processor.c: removed extra #include directives from
	the source file
	* TODO: updated TODO

2014-02-04 Jack Pugmire <comradetechnology@gmail.com>
	* CMakeLists.txt, src/CMakeLists.txt, src/compiler/makefile,
	src/machine/makefile: replaced makefiles with CMakeLists
	* INSTALL: added INSTALL
	* README.md: updated README.md
	* TODO: updated TODO
	* ChangeLog: bracketed emails with '<' and '>'

2014-02-04 Joe Jevnik <joejev@gmail.com>
	* src/machine/memory.c: fixed loadfile to not stop when reading a byte
	of value 10 or '\n'.
	* src/machine/processor.c, src/machine/processor.h,
	src/machine/operations.c, src/machine/operations.h,
	src/machine/register.h, src/compiler/compiler.c,
	src/compiler/compiler.h, src/compiler/lexer.c, src/compiler/lexer.h:
	Changed the comments at the top to be the gnu public license version.
	* src/compiler/verbose.h: fixed stdio name clash by changing our vprintf
	to v_printf.

2014-02-04 Jack Pugmire <comradetechnology@gmail.com>
	* src/machine/memory.c: made load_file less stupid
	* src/compiler/lexer.c, src/compiler/lexer.h,
	src/compiler/compiler.h, src/compiler/compiler.c: moved unsafe #include
	directives into header files
	* src/compiler/verbose.h: changed v_printf to v_msgf in order to avoid
	confusion with vprintf

2014-02-07 Joe Jevnik <joejev@gmail.com>
	* src/machine/memory.c: load_file() now sets stack_start to the next
	address after the file.
	* src/machine/processor.c: Moved functions in run() to operators.c
	* src/machine/processor.h: Added appropriate headers
	* src/machine/operations.c: Added support for all but HALT and READ
	* src/machine/operations.h: Added appropriate header
	* src/compiler/compiler.c: fixed expr_byte_len() to accuratly reflect
	the lenght of an MSET_MADDR command, added init and free of no_param
	to be shared amongst all the expr_t's
	* src/compiler/compiler.h: forward declare no_param
	* src/compiler/lexer.c: Redefined param_t to not have void *data; now
	carries a char *str that will collect error messages. Fixed issue where
	mset would segfault at lexing. Only allocate a single instance of
	no_param, free_param(no_param) == id(no_param)
	* src/compiler/lexer.h: redefined param_t
	* TODO: split TODO into MACHINE TODO and COMPILER TODO to denote the
	different actions we need to take against each component of the 16c
	system
	* examples/mset.16c: Added simple example that writes to, and reads from
	memory.
	* examples/stack.16c: Added simple example that pushes onto and pops
	from the stack.

2014-02-08 Jack Pugmire <comradetechnology@gmail.com>
	* src/machine/memory.c: changed load_file(), removing call to
	fstat()
	* src/machine/processor.h, src/machine/processor.c: added proc_tick()

2014-02-08 Joe Jevnik <joejev@gmail.com>
	* src/machine/processor.c: Changed 0xff to OP_TERM (no magic numbers)
	* ChangeLog: Fixed dates for Jack's and My last entries.

2014-02-11 Joe Jevnik <joejev@gmail.com>
	* src/common/arch.h: Added new opcodes that support refering to the
	memory address stored in a register.
	* src/compiler/compiler.c, src/compiler/compiler.h,
	src/compiler/lexer.c, src/compiler/lexer.h: Added support for the new
	suite of mset commands. Also made jump commands malloc a new param_1
	so that jumps dont share the last resolution.
	* src/machine/operations.c: Changed op_mset() to allow for the
	dereferenced registers.
	* TODO: Updated TODO.

2014-02-18 Joe Jevnik <joejev@gmail.com>
	* src/common/arch.h: Fixed type nomenclature.
	* src/machine/memory.c: Made all the memory reside in the mmaped file,
	fixed nomenclature of types.
	* src/machine/memory.h, src/machine/processor.h,
	src/machine/register.h, src/compiler/verbose.h
	src/machine/processor.h, src/machine/operations.h,
	src/machine/operations.c, src/compiler/lexer.h,
	src/compiler/lexer.c, src/compiler/compiler.h,
	src/compiler/compiler.c: Fixed type nomenclature.
	* TODO: Updated TODO.
	* src/STYLE.md: Added src/STYLE.md to discuss the style guidelines to
	follow when working on 16candles.

2014-02-18 Joe Jevnik <joejev@gmail.com>
	* src/machine/processor.c, src/machine/processor.h: Added timing to
	the processor, where each tick has a minimum time.
	* src/machine/memory.c: No longer unlink the file.
	* src/peripherals/input/input.c: Added a stdin device.
	* src/peripherals/memecho/memecho.c: Added a device that prints any
	changes to the memory, virtual and real.
	* TODO: Updated TODO.

2014-03-05 Joe Jevnik <joejev@gmail.com>
	* scr/machine/operations.c: Fixed implementation of op_mset() so that it
	correctly parses all 6 possible paramater arrangements (suffixes).
	* README.md: Added a section pertaining to additional projects and added
	h16cc to the list.

2014-03-06 Joe Jevnik <joejev@gmail.com>

src/lexer.h
/* lexer.h --- lexing functions for the 16candles compiler.
   Copyright (c) 2014 Joe Jevnik

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2 of the License, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
   more details.

   You should have received a copy of the GNU General Public License along with
   this program; if not, write to the Free Software Foundation, Inc., 51
   Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. */

#ifndef C16_LEXER_H
#define C16_LEXER_H

#include "../16common/common/arch.h"
#include "verbose.h"

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <ctype.h>
#include <assert.h>



typedef c16_halfword valtype_t;

#define INVALID_PARAM 0xff // Represents that this parameter was not parsed
                           // correctly.
#define NO_PARAM      0x00 // Represents that there is no parameter for the
                           // given parameter slot.
#define REG_PARAM     0x01 // Represents the parameter holds a register or
                           // subregister.
#define MEMADDR_PARAM 0x02 // Represents the parameter holds a memory address.
#define WORDLIT_PARAM 0x03 // Represents the parameter holds a word litteral.
#define MEMREG_PARAM  0x04 // Represents the parameter holds a memory address
                           // that is held in a register.

// The format of the parameters holding data, and a type.
typedef struct{
    c16_word  data; // The value of this parameter.
    valtype_t type; // Represents the type that the parameter is.
    char     *str;  // Represents an error with this param.
} c16_param;

// An expression to be read.
typedef struct{
    c16_opcode op;        // The instruction.
    c16_param *param_1;   // First parameter (optional - no_param).
    c16_param *param_2;   // Second Parameter (optional - no_param).
    c16_param *param_3;   // Result location (optional - no_param).
    bool       is_valid;  // Is this expression completely valid.
    char      *str_data;  // String data if needed.
} c16_expr;

// A label for jumps.
typedef struct{
    char    *name;
    c16_word loc;
} c16_label;

// The single pointer for a non existnant parameter.
extern c16_param *no_param;

// Reads the first expression out of the string.
// Like strtok, pass NULL to take the next expression out of the last string
// passed
c16_expr *get_expr(char*);

// Parse a param from a string. Identifies if it is a litteral, memory addess,
// register, subregister, or invalid type.
c16_param *parse_param(char*);

// Parses a string to a c16_word, if the string is invalid, it will return 0 and
// set invalid_word to true.
c16_word strtoword(char*);

// Reads the next parameter, this function will account for strtok return null
// prematurly.
c16_param *next_param();

// Takes a command an finds looks up the proper opcode for it or returns
// OP_INVALID if it is not a valid command.
c16_opcode parse_opcode(char*);

// Specializes a general expression.
void specialize_expr(c16_expr*);

// Checks to see if a param is valid, and adjusts the c16_expr accordingly.
void check_param(c16_expr*,c16_param*,bool);

// Frees a c16_param, freeing the data only if it holds an error string.
void free_param(c16_param*);

// Frees an c16_expr. Checks to make sure it only frees the no_param c16_param
// once.
void free_expr(c16_expr*);

#endif


src/verbose.h
/* verbose.h --- The flag and macro to deal with verbose mode
   Copyright (c) 2014 Joe Jevnik

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2 of the License, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
   more details.

   You should have received a copy of the GNU General Public License along with
   this program; if not, write to the Free Software Foundation, Inc., 51
   Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. */

#ifndef VERBOSE_MODE_H
#define VERBOSE_MODE_H

// The flag to print verbose output.
extern int is_verbose;

// Print statments that are only seen while in verbose mode.
#define v_msgf(fmt,...) if (is_verbose) { printf(fmt, ## __VA_ARGS__); }

#endif


src/compiler.h
/* compiler.h --- compilation functions for the 16candles compiler.
   Copyright (c) 2014 Joe Jevnik

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2 of the License, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
   more details.

   You should have received a copy of the GNU General Public License along with
   this program; if not, write to the Free Software Foundation, Inc., 51
   Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. */

#ifndef C16_COMPILER_H
#define C16_COMPILER_H

#include "../16common/common/arch.h"
#include "lexer.h"
#include "verbose.h"

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <getopt.h>


// The type that defines a unit of source code.
typedef struct{
    unsigned int linem;    // Line Max.
    unsigned int linec;    // Line Count.
    char       **linev;    // Line Values.
    bool         is_valid; // Is this source valid in its entirety.
    unsigned int exprm;    // Expressions Max.
    unsigned int exprc;    // Expressions Count.
    c16_expr   **exprv;    // Expressions Values.
    unsigned int bytec;    // Byte Count: Used in label resolution.
    unsigned int labelm;   // Label Max.
    unsigned int labelc;   // Label Count.
    c16_label  **labelv;   // Label Values.
} c16_src;

// An array of halfwords and the number of halfwords to write.
typedef struct{
    c16_halfword *bs;  // The expression in bytecode.
    c16_halfword  len; // The length of the bytecode.
} c16_exprbytes;

// Adds an expression to the source struct.
void add_expr(c16_src*,c16_expr*);

// Adds a label to the source label vector.
void add_label(c16_src*,char*);

// Returns true iff the c16_expr is setup in a valid way.
bool is_valid_expr(c16_expr*);

// Writes the expression to the file.
void write_expr(c16_expr*,FILE*);

// Returns the number of bytes the expression will be in bytecode.
c16_halfword expr_byte_len(c16_expr*);

// Writes the src to the file.
void write_src(c16_src*,char*);

// Allocates a new struct source struct from a file.
c16_src *src_from_file(FILE*);

// Frees a source struct.
void free_src(c16_src*);

// Attempts to compile the source to the out file. Returns the status of
// compilation
bool compile(c16_src*,char*);

#endif


src/lexer.c
/* lexer.c --- lexing functions for the 16candles compiler.
   Copyright (c) 2014 Joe Jevnik

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2 of the License, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
   more details.

   You should have received a copy of the GNU General Public License along with
   this program; if not, write to the Free Software Foundation, Inc., 51
   Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. */

#include "lexer.h"

static bool invalid_word = false;

c16_param *no_param;

// Takes the first expression from a string.
// Pass null to take the next expression out of the string.
c16_expr *get_expr(char *str){
    char *in = (str) ? str : NULL;
    c16_expr *ret = malloc(sizeof(c16_expr));
    char *tok;
    static const char* const bin_ops[][2] = { { "and",   "&&" },
                                              { "or",    "||" },
                                              { "xand",  "!&" },
                                              { "xor",   "!|" },
                                              { "lshift","<<" },
                                              { "rshift",">>" },
                                              { "add",   "+"  },
                                              { "sub",   "-"  },
                                              { "mul",   "*"  },
                                              { "div",   "/"  },
                                              { "mod",   "%"  },
                                              { "min",   ""   },
                                              { "max",   ""   } };
    static const char* const un_ops[][2] =  { { "inv",   "~"  },
                                              { "inc",   "++" },
                                              { "dec",   "--" },
                                              { "set",   "="  },
                                              { "swap",  "\\\\" },
                                              { "gt",    ">"  },
                                              { "lt",    "<"  },
                                              { "gte",   ">=" },
                                              { "lte",   "<=" },
                                              { "eq",    "==" },
                                              { "neq",   "!=" } };
    static const char* const nrp_ops[][2] = { { "push",  ":"  },
                                              { "write", ""   },
                                              { "pop",   "$"  },
                                              { "peek",  "@"  },
                                              { "read",  ""   } };
    static const char* const fr_ops[][2] =  { { "flush", "#"  },
                                              { "halt",  ""   },
                                              { "nop",   ""   } };
    static const char* const jmp_ops[][2] = { { "jmp",   "=>" },
                                              { "jmpt",  "->" },
                                              { "jmpf",  "<-" } };
    int n;
    ret->is_valid = true;
    tok = strtok(in," \n");
    if (!tok){
        ret->op       = OP_TERM;
        ret->param_1  = no_param;
        ret->param_2  = no_param;
        ret->param_3  = no_param;
        ret->str_data = NULL;
        return ret;
    }
    if (tok[0] == '@' && strlen(tok) > 1){
        ret->op       = OP_LABEL;
        ret->param_1  = no_param;
        ret->param_2  = no_param;
        ret->param_3  = no_param;
        ret->str_data = &tok[1];
        v_msgf("Adding label to expr list: '%s'\n",tok);
        return ret;
    }
    if (!strcmp(tok,"mset") || !strcmp(tok,":=")){
        ret->op = OP_MSET_;
        ret->param_1  = next_param();
        ret->param_2  = next_param();
        ret->param_3  = no_param;
        ret->str_data = NULL;
        specialize_expr(ret);
        return ret;
    }
    for (n = 0;n < 13;n++){
        if (!strcmp(tok,bin_ops[n][0])
            || !strcmp(tok,bin_ops[n][1])){
            ret->op       = parse_opcode(tok);
            ret->param_1  = next_param();
            ret->param_2  = next_param();
            ret->param_3  = next_param();
            ret->str_data = NULL;
            specialize_expr(ret);
            return ret;
        }
    }
    for (n = 0;n < 11;n++){
        if (!strcmp(tok,un_ops[n][0])
            || !strcmp(tok,un_ops[n][1])){
            ret->op       = parse_opcode(tok);
            ret->param_1  = next_param();
            ret->param_2  = next_param();
            ret->param_3  = no_param;
            ret->str_data = NULL;
            specialize_expr(ret);
            return ret;
        }
    }
    for (n = 0;n < 5;n++){
        if (!strcmp(tok,nrp_ops[n][0])
            || !strcmp(tok,nrp_ops[n][1])){
            ret->op       = parse_opcode(tok);
            ret->param_1  = next_param();
            ret->param_2  = no_param;
            ret->param_3  = no_param;
            ret->str_data = NULL;
            specialize_expr(ret);
            return ret;
        }
    }
    for (n = 0;n < 3;n++){
        if (!strcmp(tok,fr_ops[n][0])
            || !strcmp(tok,fr_ops[n][1])){
            ret->op       = parse_opcode(tok);
            ret->param_1  = no_param;
            ret->param_2  = no_param;
            ret->param_3  = no_param;
            ret->str_data = NULL;
            specialize_expr(ret);
            return ret;
        }
    }
    for (n = 0;n < 3;n++){
        if (!strcmp(tok,jmp_ops[n][0])
            || !strcmp(tok,jmp_ops[n][1])){
            ret->op            = parse_opcode(tok);
            ret->param_1       = malloc(sizeof(c16_param));
            ret->param_1->type = WORDLIT_PARAM;
            ret->param_1->data = 0;
            ret->param_2       = no_param;
            ret->param_3       = no_param;
            ret->str_data      = strdup(strtok(NULL," \n"));
            return ret;
        }
    }
    v_msgf("Read an unreckognized token: '%s'\n",tok);
    no_param = malloc(sizeof(c16_param));
    no_param->type = NO_PARAM;
    no_param->str  = NULL;
    ret->op        = OP_INVALID;
    ret->param_1   = no_param;
    ret->param_2   = no_param;
    ret->param_3   = no_param;
    ret->str_data  = strdup(tok);
    specialize_expr(ret);
    return ret;
}

void specialize_expr(c16_expr *expr){
    char *buffer;
    if (expr->op == OP_INVALID){
        expr->is_valid = false;
        return;
    }
    if (expr->op <= OP_MAX_){
        check_param(expr,expr->param_1,false);
        check_param(expr,expr->param_2,false);
        check_param(expr,expr->param_3,true);
        if (expr->param_1->type == WORDLIT_PARAM){
            if (expr->param_2->type == WORDLIT_PARAM){
                expr->op += LIT_LIT;
            }else{
                expr->op += LIT_REG;
            }
        }else{
            if (expr->param_2->type == WORDLIT_PARAM){
                expr->op += REG_LIT;
            }else{
                expr->op += REG_REG;
            }
        }
        return;
    }
    if (expr->op <= OP_LT_){
        check_param(expr,expr->param_1,false);
        check_param(expr,expr->param_2,false);
        if (expr->param_1->type == WORDLIT_PARAM){
            if (expr->param_2->type == WORDLIT_PARAM){
                expr->op += LIT_LIT;
            }else{
                expr->op += LIT_REG;
            }
        }else{
            if (expr->param_2->type == WORDLIT_PARAM){
                expr->op += REG_LIT;
            }else{
                expr->op += REG_REG;
            }
        }
        return;
    }
    if (expr->op <= OP_SET_){
        check_param(expr,expr->param_1,false);
        check_param(expr,expr->param_2,true);;
        if (expr->param_1->type == WORDLIT_PARAM){
            expr->op += LIT;
        }else{
            expr->op += REG;
        }
        return;
    }
    if (expr->op <= OP_WRITE_){
        check_param(expr,expr->param_1,false);
        if (expr->param_1->type == WORDLIT_PARAM){
            expr->op += LIT;
        }else{
            expr->op += REG;
        }
        return;
    }
    if (expr->op == OP_MSET_){
        if (expr->param_1->type == MEMADDR_PARAM){
            check_param(expr,expr->param_2,true);
            expr->op = OP_MSET_MEMADDR;
        }else if (expr->param_1->type == MEMREG_PARAM){
            check_param(expr,expr->param_2,true);
            expr->op = OP_MSET_MEMREG;
        }else{
            if (expr->param_1->type != NO_PARAM){
                if (expr->param_1->type == WORDLIT_PARAM){
                    if (expr->param_2->type == MEMADDR_PARAM){
                        expr->op = OP_MSET_LIT_MEMADDR;
                        return;
                    }else if(expr->param_2->type == MEMREG_PARAM){
                        expr->op = OP_MSET_LIT_MEMREG;
                        return;
                    }
                }else if (expr->param_1->type == REG_PARAM){
                    if (expr->param_2->type == MEMADDR_PARAM){
                        expr->op = OP_MSET_REG_MEMADDR;
                        return;
                    }else if (expr->param_2->type == MEMREG_PARAM){
                        expr->op = OP_MSET_REG_MEMREG;
                        return;
                    }
                }else{
                    expr->is_valid = false;
                }
            }
            return;
        }
    }
    if (expr->op == OP_SWAP){
        check_param(expr,expr->param_1,true);
        check_param(expr,expr->param_2,true);
        return;
    }
    if (expr->op == OP_POP || expr->op == OP_PEEK || expr->op == OP_READ){
        check_param(expr,expr->param_1,true);
    }
}

// Checks is a param is valid.
void check_param(c16_expr *expr,c16_param *param,const bool only_register){
    if ((!only_register && param->type == WORDLIT_PARAM)
        || param->type == REG_PARAM){
        return;
    }
    expr->op = OP_INVALID;
    if (param->type == NO_PARAM){
        v_msgf("Missing parameter");
        param->str = malloc(64 * sizeof(char));
        snprintf(param->str,64,"Invalid parameter: 'missing': expected "
                 "parameter of type register%s",param->data,
                 (only_register) ? "." : "or word litteral.");
        return;
    }
    v_msgf("Read invalid parameter: \"%x\"\n",param->data);
    param->type = INVALID_PARAM;
    param->str = malloc(64 * sizeof(char));
    snprintf(param->str,64,"Invalid parameter: '%s': expected "
             "parameter of type register%s",param->data,
             (only_register) ? "." : "or word litteral.");
}

// Parses a parameter out of a string.
c16_param *parse_param(char* str){
    c16_param *ret = malloc(sizeof(c16_param));
    c16_word wordlit;
    static char *rs[] = { "ipt","spt","ac1","tst","inp","ac2",
                          "r0","r1","r2","r3","r4","r5","r6","r7","r8","r9",
                          "inp_r","inp_w","r0_f","r0_b",
                          "r1_f","r1_b","r2_f","r2_b","r3_f","r3_b",
                          "r4_f","r4_b","r5_f","r5_b","r6_f","r6_b",
                          "r7_f","r7_b","r8_f","r8_b","r9_f","r9_b" };
    static c16_opreg ts[] = { OP_ipt,OP_spt,OP_ac1,OP_tst,OP_inp,OP_ac2,
                              OP_r0,OP_r1,OP_r2,OP_r3,OP_r4,OP_r5,OP_r6,OP_r7,
                              OP_r8,OP_r9,OP_inp_r,OP_inp_w,OP_r0_f,
                              OP_r0_b,OP_r1_f,OP_r1_b,OP_r2_f,OP_r2_b,OP_r3_f,
                              OP_r3_b,OP_r4_f,OP_r4_b,OP_r5_f,OP_r5_b,OP_r6_f,
                              OP_r6_b,OP_r7_f,OP_r7_b,OP_r8_f,OP_r8_b,OP_r9_f,
                              OP_r9_b };
    int n;
    if (str[0] == '*'){
        for (n = 0;n < 38;n++){
            if (!strcmp(&str[1],rs[n])){
                ret->data = ts[n];
                ret->type = MEMREG_PARAM;
                return ret;
            }
        }
        wordlit = strtoword(&str[1]);
        if (invalid_word){
            ret->type = INVALID_PARAM;
            ret->str = strdup(str);
            return ret;
        }
        ret->type = MEMADDR_PARAM;
        ret->data = wordlit;
        return ret;
    }
    for (n = 0;n < 38;n++){
        if (!strcmp(str,rs[n])){
            ret->data = ts[n];
            ret->type = REG_PARAM;
            return ret;
        }
    }
    wordlit = strtoword(str);
    if (invalid_word){
        ret->type = INVALID_PARAM;
        ret->str = strdup(str);
        return ret;
    }
    ret->type = WORDLIT_PARAM;
    ret->data = wordlit;
    return ret;
}

// Converts a string into a c16_word.
// returns 0 if it is invalid and sets invalid_word to true;
c16_word strtoword(char *str){
    invalid_word = false;
    int n;
    if (!strncmp(str,"0X",2) || !strncmp(str,"0x",2)){
        for (n = 2;n < strlen(str);n++){
            if (!isxdigit(str[n])){
                invalid_word = true;
                return 0;
            }
        }
        return (c16_word) strtoul(str,NULL,16);
    }
    for (n = 0;n < strlen(str);n++){
        if (!isdigit(str[n])){
            invalid_word = true;
            return 0;
        }
    }
    return (c16_word) strtoul(str,NULL,10);
}

// Takes the next parameter from the file, and accounts for reading EOF
// unexpectedly.
c16_param *next_param(){
    char *tok;
    c16_param *ret;
    tok = strtok(NULL," \n");
    if (!tok || tok[0] == '#'){
        ret = malloc(sizeof(c16_param));
        ret->type = INVALID_PARAM;
        ret->str = strdup("Expected additional parameter.");
        return ret;
    }
    return parse_param(tok);
}

// Detirmines the opcode of the command passed.
c16_opcode parse_opcode(char* str){
    static char *ops[][2] = { { "and",   "&&"   },
                              { "or",    "||"   },
                              { "xand",  "!&"   },
                              { "xor",   "!|"   },
                              { "inv",   "~"    },
                              { "lshift","<<"   },
                              { "rshift",">>"   },
                              { "add",   "+"    },
                              { "sub",   "-"    },
                              { "mul",   "*"    },
                              { "div",   "/"    },
                              { "mod",   "%"    },
                              { "inc",   "++"   },
                              { "dec",   "--"   },
                              { "gt",    ">"    },
                              { "lt",    "<"    },
                              { "gte",   ">="   },
                              { "lte",   "<="   },
                              { "eq",    "=="   },
                              { "neq",   "!="   },
                              { "min",   ""     },
                              { "max",   ""     },
                              { "jmp",   "=>"   },
                              { "jmpt",  "->"   },
                              { "jmpf",  "<-"   },
                              { "push",  ":"    },
                              { "pop",   "$"    },
                              { "peek",  "@"    },
                              { "flush", "#"    },
                              { "set",   "="    },
                              { "mset",  ":="   },
                              { "swap" , "\\\\" },
                              { "halt",  ""     },
                              { "nop",   ""     },
                              { "read",  ""     },
                              { "write", ""     } };
    static c16_opcode opcodes[] = { OP_AND_,
                                    OP_OR_,
                                    OP_XAND_,
                                    OP_XOR_,
                                    OP_INV_,
                                    OP_LSHIFT_,
                                    OP_RSHIFT_,
                                    OP_ADD_,
                                    OP_SUB_,
                                    OP_MUL_,
                                    OP_DIV_,
                                    OP_MOD_,
                                    OP_INC_,
                                    OP_DEC_,
                                    OP_GT_,
                                    OP_LT_,
                                    OP_GTE_,
                                    OP_LTE_,
                                    OP_EQ_,
                                    OP_NEQ_,
                                    OP_MIN_,
                                    OP_MAX_,
                                    OP_JMP,
                                    OP_JMPT,
                                    OP_JMPF,
                                    OP_PUSH_,
                                    OP_POP,
                                    OP_PEEK,
                                    OP_FLUSH,
                                    OP_SET_,
                                    OP_MSET_,
                                    OP_SWAP,
                                    OP_HALT,
                                    OP_NOP,
                                    OP_READ,
                                    OP_WRITE_ };
    int n;
    for (n = 0;n < 36;n++){
        if ((!strcmp(str,ops[n][0]) || !strcmp(str,ops[n][1]))){
            return opcodes[n];
        }
    }
    return OP_INVALID;
}


// Frees a param unless it is no_param;
void free_param(c16_param *param){
    if (param->type == NO_PARAM){
        return;
    }
    if (param->type == INVALID_PARAM){
        v_msgf("invalid param: '%s'\n",param->str);
        free(param->str);
    }
    free(param);
}

// Frees an expression by verifying it only frees no_param once.
void free_expr(c16_expr *expr){
    free_param(expr->param_1);
    free_param(expr->param_2);
    free_param(expr->param_3);
    free(expr);
}
